### ISSUE TITLE
Common implementation of RFC 2617 digest authentication

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3575]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)| @washort reported|
|-|-|
|Trac ID|trac#3575|
|Type|enhancement|
|Created|2008-12-12 16:13:18Z|
|Branch|https://github.com/twisted/twisted/tree/common-digest-3575|

Both HTTP and SIP use digest auth, so an implementation in Twisted core that both can use would be handy.

Attachments:

* [sasl_mechanisms.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f88/f88f86327f618ec0250d696b8eacb6d5bb8ae2a3/71b978e29070dce4ec13f4313af7a60220a1c49a.patch) (1757 bytes) - added by ralphm on 2008-12-16 15:32:59Z - Patch to use common digest functions.

<details><summary>Searchable metadata</summary>

```
trac-id__3575 3575
type__enhancement enhancement
reporter__washort washort
priority__highest highest
milestone__ 
branch__branches_common_digest_3575 branches/common-digest-3575
branch_author__washort washort
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1229098398000000 1229098398000000
changetime__1231446420000000 1231446420000000
version__None None
owner__washort washort
cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_1"></a>|@washort commented|
|-|-|

(In [25639]) Branching to 'common-digest-3575'

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)|@washort removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm set owner to @washort|
|-|-|

This ticket is a partial dupe of [[#2015](https://github.com/twisted/twisted/issues/2015)](https://github.com/twisted/twisted/issues/2015). The SASL code for Jabber also implements its own Digest code, and should be included in this patch at the very least. But I'd rather see progress on #2015.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

This ticket's easier to resolve than [[#2015](https://github.com/twisted/twisted/issues/2015)](https://github.com/twisted/twisted/issues/2015), though, and #2015 will benefit from its resolution (by no longer having to concern itself with the precise details of the crypto for digest auth).

Also, smaller branches are better.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @washort<br>@exarkun set status to `new`|
|-|-|

  1. twisted/web/iweb.py - removing `IUsernameDigestHash` from here is backward incompatible.  The name should still be exposed here, which is fortunately easy, just import it from `twisted.cred.digest`.
  1. twisted/cred/digest.py
    1. wrong test-case-name
    1. when implemented in `twisted.web`, most of the digest implementation was private.  Now it's all public.  It'd be better if it remained private, at least for now.
  1. twisted/test/test_digestauth.py - ()-style imports aren't python 2.3 compatible
  1. twisted/protocols/sip.py
    1. several unused imports - random, sys, md5
    1. The removals are all incompatible changes.  The functions (`DigestCalcHA1`, `DigestCalcResponse`), interface, and classes should be left in place and deprecated.
    1. The change to `RegisterProxy.authorizers` is incompatible.  It used to be a mapping of names to `IAuthorizer` providers, now its values are something else.  This will break any code that is still trying to insert `IAuthorizer` providers.  I suggest writing an adapter from `IAuthorizer` to `ICredentialFactory` to support both interfaces.  The adapter might emit a deprecation warning any time it is invoked (this is probably easier than trying to get a deprecation emitted any time someone `implements(IAuthorizer)` anyway).
  1. twisted/test/test_sip.py - please update copyright date
  1. there's still a tiny property change on /

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_9"></a>|@washort removed owner|
|-|-|

Fixed. I'm not completely happy with the way the compatibility stuff came out for SIP, but let me know what you think.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_10"></a>|@ralphm set owner to @washort|
|-|-|

Review comments:

 * In the shuffle to make some stuff private, you missed fixing `twisted/web/_auth/digest.py`. `DigestCredentialFactory` is now in `twisted.cred.credentials`
 * I notice that most of this code is server-side oriented. However, the hashing functions in (now) `twisted.cred._digest` are also useful for the client side. I have modified `t.w.p.j.sasl_mechanism.DigestMD5._gen_response` to use them. See the patch file. Might be good to include in this branch.
 * Given the above, `twisted.cred._digest` should probably _not_ be private.


Observations:

 * The [diff:branches/generic-sasl-2015-2@21380:21930 branch] for [#2015](https://github.com/twisted/twisted/issues/2015) has very similar functions in its [[branches/generic-sasl-2015-2/twisted/cred/digest.py@21930](branches/generic-sasl-2015-2/twisted/cred/digest.py@21930) `twisted.cred.digest`].
 * The latter implementation supports some SASL specifics (RFC:2831), among which an additional parameter to `calcHA1`: `pszAuthzID`. My SASL code doesn't use that yet, but it would be nice if that were supported.
 * The decoding of a response is very crude. In some scenarios, quoted values may have comma's inside. I don't believe this is something you will run into if you only implement the server-side of HTTP-only DIGEST MD5, but this is an accident waiting to happen. See also [[branches/generic-sasl-2015-2/twisted/cred/digest.py@21930#L41](branches/generic-sasl-2015-2/twisted/cred/digest.py@21930#L41) `_parseDigest`] method the branch for [#2015](https://github.com/twisted/twisted/issues/2015).

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_11"></a>|@washort removed owner|
|-|-|

Fixed the import.

Changes to the digest code itself I think are outside the scope of this branch -- my goal is primarily to deprecate the digest auth code in `twisted.protocols.sip`. Making other auth methods share code with this HTTP Digest stuff can be done after this branch is merged, whether in [#2015](https://github.com/twisted/twisted/issues/2015) or a new ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @washort|
|-|-|

Failing unit tests

```
===============================================================================
[ERROR]: twisted.web2.test.test_httpauth

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/runner.py", line 555, in loadPackage
    module = modinfo.load()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/modules.py", line 380, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/python/reflect.py", line 462, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/web2/test/test_httpauth.py", line 8, in <module>
    from twisted.cred.digest import calcResponse, calcHA1, calcHA2
exceptions.ImportError: No module named digest
-------------------------------------------------------------------------------
```

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_13"></a>|@washort removed owner|
|-|-|

Whoops, fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @washort<br>@exarkun set status to `new`|
|-|-|

  1. regarding the attached patch to make `twisted.words` use this code, please file a new ticket for that.  I hope that the addition of a higher-level public API for the client side of this story will remove the need to import private APIs into `twisted.words` (as the patch has to do now).  Cred hasn't paid a lot of attention to client concerns thus far, and the code being refactored in this branch is all server-code.
  1. twisted/web/iweb.py - `__all__` is supposed to contain strings
  1. twisted/protocols/sip.py
    1. The next release of Twisted will be 8.3; all the deprecations should refer to that version.
    1. Deprecations (`DigestCalcHA1`, `DigestCalcResponse`, `_AuthorizerAdapter`, `BasicAuthorizer`, `DigestedCredentials`, `DigestAuthorizer`) need unit tests.  See `TestCase.callDeprecated` or `TestCase.assertWarns` or `TestCase.flushWarnings`.
    1. `_AuthorizerAdapter` is entirely untested and its `getChallenge` method is buggy.
    1. `BasicAuthorizer` is entirely untested.
    1. The exception codepaths in `RegisterProxy.login` and the adaption failure codepath are not tested.
    1. `RegisterProxy._ebLogin` is untested
  1. twisted/test/test_sip.py
    1. it'd be great if this module didn't need to import private stuff from `twisted.cred`.  There's not really any new digest code in `twisted.protocols.sip`, so maybe the tests don't have to do real digest.  What if the tests which are currently using those private APIs (just `AuthorizationTestCase.testChallenge`, I think) were changed to do non-digest auth?  That should exercise the `ISIPAuthorizer`/`RegisterProxy` interaction well enough.  Some separate unit tests for `_DCFWrapper` could test _it_ against a `DigestCredentialFactory` which behaves deterministically (as `FakeDigestCredentialFactory` does) and could just hardcode the correct response rather than computing it (it might even use the samples from the digest auth rfc and say so).  `_DCFWrapper` should have more direct unit tests anyway.
    1. `AuthorizationTestCase` could use some docstrings so it's clear which unit it's trying to test (it's not very unity at the moment, so maybe you'll have to figure out what unit it _should_ be testing is and adjust it a bit - see above about my guess: `ISIPAuthorizer`/`RegisterProxy` interaction).
    1. `_DCFWrapper` might be easier to test if it accepted a `DigestCredentialFactory` instance instead of arguments for creating one
  1. twisted/web/_auth/digest.py has trailing whitespace
  1. twisted/web/test/test_httpauth.py has trailing whitespace
  1. twisted/web2/ - probably shouldn't make these changes.  They might break someone's application, and the general code quality of web2 isn't a major priority.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_16"></a>|@washort removed owner|
|-|-|

Upon reflection, I think I'm not going to change the existing SIP code to use the new API, since all the digest-auth-using code in `twisted.protocols.sip` needs to be deprecated too. So for now I'll just deprecate the SIP digest bits, and use the new API for new SIP code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @washort|
|-|-|

  1. From the previous review, points 1. and 2. remain unaddressed.
  1. From ralphm's review, the third observation should be tracked somehow.  Please file a ticket for making `DigestCredentialFactory.decode` parse response strings correctly according to the specification.
  1. Please make the `generateNonce`, `generateOpaque`, and `verifyOpaque` methods of `DigestCredentialFactory` private

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_18"></a>|@washort set owner to @exarkun|
|-|-|

Done.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to @washort|
|-|-|

  1. Update copyright date on all modified files to include 2009
  1. The hashlib import in `twisted/protocols/sip.py` should be in the Twisted imports section
  1. `twisted/web/iweb.py` gained `IRequest` in the mean time.  Can you update `__all__` to include that as well?

All trivial points, please merge when they're addressed.

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_20"></a>|@washort set status to `closed`|
|-|-|

(In [25898]) Merge common-digest-3575.
Move digest-auth code out of twisted.web and into twisted.cred.
Author: washort
Reviewer: exarkun
Closes [#3575](https://github.com/twisted/twisted/issues/3575).

**Comment by twisted-trac** [comment]
|[<img alt="washort's avatar" src="https://avatars.githubusercontent.com/u/947487?s=50" width="50" height="50">](https://github.com/washort)<a name="note_21"></a>|@washort commented|
|-|-|

(In [25902]) Merge common-digest-3575.
Move digest-auth code out of twisted.web and into twisted.cred.
Author: washort
Reviewer: exarkun
Closes [#3575](https://github.com/twisted/twisted/issues/3575).
